Add Backward Compatibility Testing documentation (#10)
vmware-tanzu/tanzu-plugin-runtime: test/CROSS_VERSION_API_COMPATIBILITY.md

As a test writer you should be knowledgeable on specific APIs and specific Runtime libraries i.e. what arguments each API takes and what it returns.
Test writers should also be familiar with how to write combination tests with multiple APIs i.e. when testing mutators(Set***, Delete*** APIs) and reader(Get*** APIs) together.
When writing a test involving two versions of Runtime library APIs a test writer should be aware of what specific fields are added or removed from the API method for those 2 versions and design tests accordingly.

### Commands

Run all `****_test.go` files from `compatibility-tests` directory.

``` shell
make compatibility-tests
```

### How are the Tests being run ?

The tests are executed as a GitHub runner CI pipeline that executes the below command.

``` shell
make compatibility-tests
```

### What will the tests cover ?

Interactions of different implementations of APIs across various versions of library works as expected.

### What will be the test results ?

GitHub CI runner pipeline include Logs with details on test cases from test-case that are executed and succeeded/failed.

## API and Test cases details

### Test Suite API

> Framework has `framework.NewTestCase` to write a specific test case, each test case consist sequence of commands, each command is specific to a runtime api call.

Reviewer: ```suggestion
Framework has `framework.NewTestCase` to write a specific test case, each test case consists of a sequence of commands, with each command corresponding to an invocatino of a API from a specific version of runtime.
```

> We need to use the above high level APIs to construct specific test command, the input data for each high level apis changes based on runtime version's its being tested, for more details look into the input parameter's type definition.

Reviewer: high level APIs -> do you mean the test helper functions? The only API mentioned so far are plugin runtime APIs. If you intend to refer to the framework test APIs, perhaps indicate clearly like in L51.

Author: updated and moved to High level test API section

``` go
// Construct series of commands to execute
 testCase := framework.NewTestCase().Add(setContextCommand).Add(getContextCommand)
```

### High-level test API

> To write compatibility test cases, the test writer should be aware of below high level framework Test APIs, each below api helps to test specific runtime api by constructing test command:

Reviewer: there are still api/API inconsistencies are various parts of the docs

Author: addressed all now

``` go
func NewSetXXXCommand(setXXXInputOptions SetXXXInputOptions, setXXXOutputOptions SetXXXOutputOptions) (*Command, error)
func NewGetXXXCommand(getXXXInputOptions GetXXXInputOptions, getXXXOutputOptions GetXXXOutputOptions) (*Command, error)
```

``` go

// NewSetContextCommand contructs a command to make a call to specific runtime version SetContext API.
// Input Parameter: setContextInputOptions has all input parameters which are required for Runtime SetContext API.
// Input Parameter: setContextOutputOptions has details about expected output from Runtime SetContext API call.

> // Return:  command to execute or error if any validations fails for SetContextInputOptions or SetContextOutputOptions

Reviewer: ```suggestion
// Return: command to execute or error if any validations fails for SetContextInputOptions or SetContextOutputOptions
```

> // This method does validates the input parameters  SetContextInputOptions/SetContextOutputOptions based on Runtime API Version

Reviewer: missing a word or two here near the double space gap?

Author: fixed

```

More details about input parameters for above high level api:

``` go
const (
  Version0110 RuntimeVersion = "v0.11.0"
  Version0250 RuntimeVersion = "v0.25.0"
  Version0280 RuntimeVersion = "v0.28.0"
  Version100 RuntimeVersion = "v1.0.0"
)

type RuntimeVersion string

type RuntimeAPIVersion struct{
 RuntimeVersion RuntimeVersion
}

type GetContextInputOptions struct {
 RuntimeAPIVersion // required
 ContextName string // required
}

type GetContextOutputOptions struct{
 RuntimeAPIVersion // required
 CtxOptions // For specific version options look into CtxOptions definition
}

type SetContextInputOptions struct{
 RuntimeAPIVersion // required
 CtxOptions // required
}

type SetContextOutputOptions struct{
 error string // expected error message could be the sub string of actual error message
}

type DeleteContextInputOptions struct{
 RuntimeAPIVersion // required
 ContextName string // required
}

type DeleteContextOutputOptions struct{
 error string // expected error message could be the sub string of actual error message
}
```

### xxxOptions structs

> - Each of these struct represents the super set of all fields for their respective config types (e.g .Context, Server, DiscoverySource) across for all Runtime versions in which the types are defined.

Reviewer: ```suggestion
- Each of these structs represents the super set of all fields for their respective config types (e.g .Context, Server, DiscoverySource) across all Runtime versions in which the types are defined.
```

> - Based on the Runtime Version xxxOptions attributes may change(mandatory/optional/Not applicable).

Reviewer: ```suggestion
- Based on the Runtime Version xxxOptions attributes may change (mandatory/optional/Not applicable).
```

Reviewer: combining L121-122:
The xxxOptions structs are used as inputs to the API invocation as well as data to verify against the output of the invocation.The implications of setting or not setting certain attributes changes depending which scenario under which it is being used.

Author: updated

- It is meant to be used as both inputOptions and outputOptions.
- Command Creation Helper functions (i.e. NewSetContextCommand) validate the supplied inputOptions as per RuntimeVersion and make sure all required attributes are set. If not Command Creation Helper functions will throw error and Test fails.

> - The validation of “Input and expect output” happens in  Set/Get/DeleteXXXCommand creation (within framework.NewTestCase), so it's not part of test case execution. so its kind of test case setup.

Reviewer: ```suggestion
- The validation of “Input and expect output” happens in Set/Get/DeleteXXXCommand creation (within framework.NewTestCase), so it's not part of test case execution. so its kind of test case setup.
```

  (Eg: GetContextOutputOptions/SetContextOutputOptions/DeleteContextOutputOptions) as its test writer responsibility, the Output data is depends on the Input data (eg: if input is not valid, runtime api may return error) OR the sequence of calls made before the current call (eg: NewGetContextCommand() output depends on the previous SetContext** calls).
  If test writer is not passing valid expected output data, then it fails during the TEST CASE EXECUTION as test case fail.
- Test framework does validate the expected Output struct’s field values with runtime api response to its entirety.
- Test writer does not have to explicitly specify to check whether an attribute value is nil as all fields default values are nil or “”.
- Test writer will need to specify what exact object attribute values he expects out of GetXX api method (framework perform a full equality check and not partial equality, means if api returns attribute which not specified by test writer then test case will fail).
- So for the output struct, all filed are optional, there are NO MANDATORY fields!!! its test writer responsibility to identify the expected data!!

> Example 1: For the Below NewSetContextCommand if the context arguments are passed incorrect

Reviewer: Please be consistent with casing of the words. Any reason why Below is capitalized.

``` go
// Input Parameters for Runtime SetContext API
   setContextInputOptions := SetContextInputOptions{
       RuntimeVersion: framework.Version0280,
       CtxOptions: CtxOptions{
           Name: "context-one",
           Type: framework.CtxTypeK8s, // Invalid attribute since Type is not supported in v0.28.0
           GlobalOpts: &framework.GlobalServer{
             Endpoint: "test-endpoint",
          }
       }
   }
   // Create SetContext Command
   setContextCommand, err := framework.NewSetContextCommand(setContextInputOptions, setContextOutputOptions)
   Expect(err).To(BeNil()) // Error is thrown since Type is not supported in specified runtime version v0.28.0

Below Error is thrown when the test case with above command input options is run
    Expected
        <*errors.fundamental | 0x140000b8030>: {
            msg: "invalid set context input options for the specified runtime version v0.28.0",
            stack: [0x1051f2ac9, 0x1051f1bf0, 0x1051f4544, 0x10500a120, 0x105009c64, 0x105009474, 0x10500c478, 0x10500bec8, 0x1050285c0, 0x10502833c, 0x105027ad0, 0x105029b5c, 0x105034da0, 0x105034c44, 0x1051f4d38, 0x104e27fd0, 0x104d91814],
        }
    to be nil
```

> Example 2  Explains on how framework validates the null fields and What is expected of Test Writer

Reviewer: ```suggestion
Example 2  How framework validates the null fields and what is expected of test writer
```
